shotover-proxy/src/transforms/redis/sink_cluster.rs

```diff
@@ -140,48 +141,52 @@ impl RedisSinkCluster {
                 };
                 Box::pin(one_rx)
             }
+            // Send to the single channel and return its response.
             1 => {
                 let channel = channels.get(0).unwrap();
                 let one_rx = self.choose_and_send(channel, message).await?;
                 Box::pin(one_rx.map_err(|_| anyhow!("no response from single channel")))
             }
+            // Send to all senders.
+            // If any of the responses were a failure then return that failure.
+            // Otherwise return the first successful result
             _ => {
                 let responses = FuturesUnordered::new();
 
                 for channel in channels {
                     responses.push(self.choose_and_send(&channel, message.clone()).await?);
                 }
 
-                // Reassemble upstream responses in any order into an array, as the downstream response.
-                // TODO: Improve error messages within the reassembled response. E.g. which channel failed.
-
                 Box::pin(async move {
                     let response = responses
-                        .fold(vec![], |mut acc, response| async move {
-                            match response {
-                                Ok(Response {
-                                    response: Ok(mut messages),
-                                    ..
-                                }) => acc.push(messages.pop().map_or(
-                                    RedisFrame::Null,
-                                    |mut message| match message.frame().unwrap() {
-                                        Frame::Redis(frame) => frame.take(),
-                                        _ => unreachable!(),
-                                    },
-                                )),
-                                Ok(Response {
-                                    response: Err(e), ..
-                                }) => acc.push(RedisFrame::Error(e.to_string().into())),
-                                Err(e) => acc.push(RedisFrame::Error(e.to_string().into())),
+                        .fold(None, |acc, response| async move {
+                            if let Some(RedisFrame::Error(_)) = acc {
+                                acc
+                            } else {
+                                match response {
+                                    Ok(Response {
+                                        response: Ok(mut messages),
+                                        ..
+                                    }) => Some(messages.pop().map_or(
+                                        RedisFrame::Null,
+                                        |mut message| match message.frame().unwrap() {
+                                            Frame::Redis(frame) => frame.take(),
+                                            _ => unreachable!(),
```

Comment:
Can we add a message explaining why this is unreachable?

Reply:
lol, its not unreachable.
While its bad code, its existing bad code.

But I think we are better off leaving it as is until we start resolving https://github.com/shotover/shotover-proxy/issues/520

Reply:
Ah, that is confusing! Sounds good to me.

Reply:
When would we get a non-Redis frame as a response here? i.e. how exactly do we 'reach' here? 🤔

Reply:
User hooks up a CassandraSource to a RedisSink

Reply:
Ah!
Thankyou for thinking about this properly
You are 100% correct.

I added a proper message to the unreachable explaining this.
It is actually unreachable because the codec will only ever return a Redis message.